server/sonar-server/src/main/java/org/sonar/server/computation/component/PathAwareVisitorAdapter.java

```diff
@@ -0,0 +1,212 @@
+      default:
+        return factory.createForUnknown(component);
+    }
+  }
+
+  public void visitProject(Component project, Path<T> path) {
+    // empty implementation, meant to be override at will by subclasses
+  }
+
+  public void visitModule(Component module, Path<T> path) {
+    // empty implementation, meant to be override at will by subclasses
+  }
+
+  public void visitDirectory(Component directory, Path<T> path) {
+    // empty implementation, meant to be override at will by subclasses
+  }
+
+  public void visitFile(Component file, Path<T> path) {
+    // empty implementation, meant to be override at will by subclasses
+  }
+
+  public void visitUnknown(Component unknownComponent, Path<T> path) {
+    // empty implementation, meant to be override at will by subclasses
+  }
+
+  public void visitAny(Component component, Path<T> path) {
+    // empty implementation, meant to be override at will by subclasses
+  }
+
+  /**
+   * A Simple implementation which uses the same factory method for all types which can be implemented by subclasses:
+   * {@link #createForAny(Component)}.
+   */
+  public abstract static class SimpleStackElementFactory<T> implements StackElementFactory<T> {
+
+    public abstract T createForAny(Component component);
+
+    @Override
+    public T createForProject(Component project) {
+      return createForAny(project);
+    }
+
+    @Override
+    public T createForModule(Component module) {
+      return createForAny(module);
+    }
+
+    @Override
+    public T createForDirectory(Component directory) {
+      return createForAny(directory);
+    }
+
+    @Override
+    public T createForFile(Component file) {
+      return createForAny(file);
+    }
+
+    @Override
+    public T createForUnknown(Component file) {
+      return createForAny(file);
+    }
+  }
+
+  public static class DequeBasedPathAdapter<T> implements DequeBasedPath<T> {
+    private final Deque<PathElement<T>> deque = new ArrayDeque<>();
+
+    @Override
+    public T current() {
+      return deque.getFirst().getElement();
+    }
+
+    @Override
+    public T parent() {
+      Iterator<PathElement<T>> iterator = deque.iterator();
+      if (iterator.hasNext()) {
+        iterator.next();
+        if (iterator.hasNext()) {
+          return iterator.next().getElement();
+        }
+      }
+      throw new NoSuchElementException("Path is either empty or has only one element. There is no parent");
+    }
+
+    @Override
+    public boolean isRoot() {
+      return deque.size() == 1;
+    }
+
+    @Override
+    public T root() {
+      return deque.getLast().getElement();
+    }
+
+    @Override
+    public Iterator<PathElement<T>> iterator() {
+      return deque.iterator();
+    }
+
+    @Override
+    public Iterable<PathElement<T>> getCurrentPath() {
+      return this;
+    }
+
+    public void add(PathElement<T> pathElement) {
+      deque.addFirst(pathElement);
+    }
+
+    public PathElement<T> pop() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add the "@Override" annotation above this method signature [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)